core/trino-main/src/main/java/io/trino/operator/function/StreamTableFunctionInput.java

```diff
@@ -0,0 +1,125 @@
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.concurrent.MoreFutures.toListenableFuture;
+import static io.trino.operator.WorkProcessorUtils.yieldingIteratorFrom;
+import static java.util.Objects.requireNonNull;
+
+public class StreamTableFunctionInput
+        implements WorkProcessor.Process<WorkProcessor<Page>>
+{
+    private final TableFunctionDataProcessor tableFunction;
+    private final int properChannelsCount;
+    private final int passThroughSourcesCount;
+    private final List<List<Integer>> requiredChannels;
+    private final Optional<Map<Integer, Integer>> markerChannels;
+    private final List<RegularTableFunctionPartition.PassThroughColumnSpecification> passThroughSpecifications;
+
+    private final Iterator<Optional<Page>> inputPages;
+    private final PagesIndex pagesIndex;
+    private boolean finished;
+
+    public StreamTableFunctionInput(
+            TableFunctionDataProcessor tableFunction,
+            int properChannelsCount,
+            int passThroughSourcesCount,
+            List<List<Integer>> requiredChannels,
+            Optional<Map<Integer, Integer>> markerChannels,
+            List<RegularTableFunctionPartition.PassThroughColumnSpecification> passThroughSpecifications,
+            WorkProcessor<Page> inputPages,
+            PagesIndex pagesIndex)
+    {
+        this.tableFunction = requireNonNull(tableFunction, "tableFunction is null");
+        this.properChannelsCount = properChannelsCount;
+        this.passThroughSourcesCount = passThroughSourcesCount;
+        this.requiredChannels = requiredChannels.stream()
+                .map(ImmutableList::copyOf)
+                .collect(toImmutableList());
+        this.markerChannels = markerChannels.map(ImmutableMap::copyOf);
+        this.passThroughSpecifications = ImmutableList.copyOf(requireNonNull(passThroughSpecifications, "passThroughSpecifications is null"));
+        this.inputPages = yieldingIteratorFrom(requireNonNull(inputPages, "inputPages is null"));
+        this.pagesIndex = requireNonNull(pagesIndex, "pagesIndex is null");
+    }
+
+    @Override
+    public ProcessState<WorkProcessor<Page>> process()
+    {
+        if (finished) {
+            return ProcessState.finished();
+        }
+
+        if (inputPages.hasNext()) {
+            Optional<Page> next = inputPages.next();
+            if (next.isEmpty()) {
+                return ProcessState.yielded();
+            }
+            Page currentInputPage = next.get();
+            pagesIndex.clear();
+            pagesIndex.addPage(currentInputPage);
+            return ProcessState.ofResult(new RegularTableFunctionPartition(
+                    pagesIndex,
+                    0,
+                    pagesIndex.getPositionCount(),
+                    new TableFunctionDataProcessor()
```

Comment:
Do we need to wrap in a new anonymous class?

Locally, I directly passing the `tableFunction` which also resolved the `TestJsonTable` failures for me.